Add sites from the site manager

bgrgicak opened the pull request:
## Motivation for the change, related issues

The current site manager allows users to switch between sites only, but it still isn't clear how to add a site.

To allow users to add sites this PR adds an _Add site_ button and modal for inserting the site name.
On submit a new site is added and loaded.

## Implementation details

The `add-site-button` component validates the name and generates a slug out of it.
Because we can only store the slug until site storage is completed, the name is limited to lowercase letters and numbers.

The site sidebar now supports adding sites. Sites are added to the state which will later get moved to site storage.

The `add-site-button` component isn't completed, this is only a temporary implementation.
Once we have the site manager, the button will just create a new site without any modals or inputs.

## Testing Instructions (or ideally a Blueprint)

- open [Playground with browser storage](http://127.0.0.1:5400/website-server/?storage=browser)
- open the site manager menu (upper left corner)
- see the _Add site_ button
- click on the button and see the modal
- add a site using the modal


![Screenshot 2024-08-09 at 06 57 28](https://github.com/user-attachments/assets/207c416f-af33-4674-b998-a2918dce49db)

bgrgicak commented on packages/playground/website/src/components/site-manager/site-manager-sidebar/index.tsx:
```diff
@@ -59,19 +75,32 @@ export function SiteManagerSidebar({
+		 * If the site name already exists, we won't need to add it again.
+		 * TODO: remove this check when site storage is implemented.
+		 * In site storage we won't be limited to having unique site names.
+		 */
```
@brandonpayton I made some assumptions here about how site storage will work but it's not necessary to implement it in that way.

bgrgicak commented on packages/playground/website/src/components/site-manager/add-site-button/style.module.css:
```diff
@@ -0,0 +1,23 @@
+	width: 100%;
+}
+
+.add-site-input {
+	margin-bottom: 24px;
+}
+.add-site-input.invalid-input :global(.components-input-control__backdrop) {
+	border-color: #e26f56 !important;
+}
+
+.add-site-input.invalid-input :global(.components-base-control__help) {
+	color: #e26f56;
+}
```
I wasn't able to find any validation in WordPress components.

bgrgicak commented on packages/playground/website/src/components/site-manager/site-manager-sidebar/index.tsx:
```diff
@@ -111,7 +140,16 @@ export function SiteManagerSidebar({
 					{sites.map((site) => {
-						const isSelected = site.slug === siteSlug;
+						/**
+						 * The default site is selected when no site slug provided.
+						 *
+						 * The default site is stored in the `wordpress` directory
+						 * and that's used as the slug.
```
```suggestion
						 * The `wordpress` site is selected when no site slug is provided.
```

bgrgicak commented on packages/playground/website/src/components/site-manager/site-manager-sidebar/index.tsx:
```diff
@@ -111,7 +140,13 @@ export function SiteManagerSidebar({
 				<MenuGroup className={css.siteManagerSidebarList}>
 					{sites.map((site) => {
-						const isSelected = site.slug === siteSlug;
+						/**
+						 * The `wordpress` site is selected when no site slug is provided.
+						 */
+						const isSelected =
+							site.slug === siteSlug ||
+							(siteSlug === undefined &&
+								site.slug === 'wordpress');
```
This is an unrelated bug fix.

adamziel commented:
The hamburger menu seems broken:

https://github.com/user-attachments/assets/aa9fcc62-8b25-4a6e-bee2-7c490136c009

adamziel commented:
I was wondering what causes the issue and couldn't crack it. I checked out trunk and this PR again and that problem is now gone. How weird!

adamziel commented:
Is there a reason we have to ask the user to invent a name? Could we just call it "My new site" and skip the modal? On the other hand, modal will have more options in the future so maybe it's a good start.

A few UX notes about the modal itself:

* The "Add site" button is inactive even after I provide a name. I assume that's because I have an uppercase letter in the text field. There's no indication of an error, though.
* Let's just accept all the characters and skip the "Can be lowercase letters, numbers, and spaces" part. If there's a technical reason for not accepting uppercase letters, let's resolve that reason and accept them. 
* Let's get rid of the 1password widget
* Pressing "enter" reloads the entire page without creating a new site

![CleanShot 2024-08-09 at 11 23 20@2x](https://github.com/user-attachments/assets/8b14c0ff-72ac-441e-89bd-0c2c522efe57)

bgrgicak commented:
> Is there a reason we have to ask the user to invent a name? Could we just call it "My new site" and skip the modal? On the other hand, modal will have more options in the future so maybe it's a good start.

There is no way to change the name today, so I wanted to give users a choice. 

> Let's just accept all the characters and skip the "Can be lowercase letters, numbers, and spaces" part. If there's a technical reason for not accepting uppercase letters, let's resolve that reason and accept them.

It's now changing, so the name can be anything. I also added validation to avoid duplicate names.
TBH it feels like an overcomplicated temporary feature. 

> Let's get rid of the 1password widget

I added `data-1p-ignore` but I can't test it. Please let me know if it worked.

> Pressing "enter" reloads the entire page without creating a new site

Fixed.

adamziel commented:
> TBH it feels like an overcomplicated temporary feature.

Agreed, it should be fine to just write that name to a `name.json` file for now so that it has no relationship with site slug whatsoever.

bgrgicak commented:
Thanks for the review @adamziel! With auto-generating names this PR should work ok for users until we get the site storage.
